rfcs/text/0000-multi-data-collection-dataset-processing.md

```diff
@@ -0,0 +1,115 @@
+
+## Detailed Design
+A new concept of *data group* is added to the DCP data model to address these issues.  A data group is defined as a set of specific versions of metadata and data that is complete and consistent by specified criteria.  A *data group* is not created until all its contents are complete.  Events are generated when a *data group* is created, updated, or deleted.
+
+A data group has a symbolic scope type that specifies what is represented by the group, as well a the set of criteria by which it is complete.  Most of the above use cases will require a *project submission* scope that indicates submission to a project is complete.
+
+Data groups are implemented as a new bundle type that contains a JSON file listing the FQIDs (UUIDs with versions) of all bundles that part of the data groups and the existing DSS subscription mechanism for notifications.
+
+A new schema *data_group* will be created that contains the fields:
+- *scope* - Symbolic name of the scope, for example, *PROJECT_SUBMISSION*.
+- *bundle_fqids* - List of FQIDs of bundles that are in the group.
+
+
+Initially, only the *PROJECT_SUBMISSION* scope will be implemented. *PROJECT_SUBMISSION* will indicate that the entirety of single project submission is version-complete and will be generated by ingesting on after all data bundles are committed to the DSS.  This is indicated that data in the project submission should be processed by downstream components.  Ingest is responsible for creating *PROJECT_SUBMISSION* *data groups* after the relevant bundles have been created.  Figure 1 shows a diagram of a project with multiple *data groups*.
+
+![Figure 1](/rfcs/images/0000-multi-data-collection-data-processing-images/project-submission-in-project.png)
+
+
+## Application to analysis pipelines
+
+While this RFC proposes a general mechanism, it was developed in response to the needs of analysis to co-process data that are in multiple assay bundles.  This section uses co-processing of multiple sequencing run from a given library as an example of the use of data groups to identify data that should be processed together.  A key concept behind this proposal is that upstream submission should not be defining how downstream analysis groups input for processing.  The analysis pipelines need to be able to group input based on ways that may not be defined at submission time.  New analysis pipelines should be able to group inputs in arbitrary ways without requiring changes to the way data is packaged.
+
+### Identifying data to co-process
+Pipelines that require grouping data from multiple sequencing assays (co-processing) subscribe to *PROJECT_SUBMISSIONS* events instead of per-bundle events. Unlike with the assay bundle per pipeline run, this makes the pipeline framework responsible for collecting assay bundles into pipeline runs.  Pipelines that do not require co-processing continue to subscribe to assay bundle events and are unaffected.
```

Comment:
Given the general possibility of top-up sequencing, isn't _every_ pipeline a project pipeline?

Reply:
good point @mckinsel 

@justincc, how is top-up sequencing handled?

Reply:
Top up would require an update to the primary data bundle to add additional fastqs or it'd be what Marcus says, every pipeline would be a project pipeline and would get a notification that new data has arrived for that sample and has to be re-run with all the data in mind. This does present a complex graph where old analysis results are for one set of inputs and new analysis results are for an aggregate.

Reply:
@kbergin So these are new FASTQ files that should be added to an existing process that produces cell_suspension -> FASTQs? In a library entity world, those FASTQs would come from the existing library entity? If so, this is indeed adding files to an existing bundle. How common is this case? From other conversations I thought adding files to exitsing bundles was a low-event/low importance case. If this isn't so then we may need to re-evaluate implementation priorities.

Reply:
I could be interpreting top-up sequencing in a different way than @mckinsel intended.

At the Broad, a top up sequencing is used to hit sequencing coverage on a sample before the data is delivered to the customer that paid for it. In our world, the customer only ever receives one aggregated BAM that has already been processed and contains all of the data from multiple sequencing events in one BAM. If they were to revert this to an unaligned raw FASTQ, it would be one submission event from HCA's perspective.

If a data submitter has data for a sample and submits it and then receives more data for the same sample, it could be a few different scenarios. In general, it implies more sequencing of the same library preparation. In my experience in the lab, sometimes a new library preparation from the same sample is created to perform the additional sequencing to top up. I'm not sure if that's applicable to droplet based single cell experiments cc @barkasn 

I imagine that for the forseeable future, as we are receiving data from the data submitters after they have likely already performed processing on it from their sequencing core or group, we won't receive as much top up sequencing after the fact. It's probably more likely that we'll have a submitter just forget to include a fastq in their initial data submission that was there from the beginning and was not a top up that arrived later.

Does that explanation help or confuse further?

Reply:
Interesting, that is different than my experience with Broad for TCGA where the goal was to get the data out quickly.  They call this BAM top-off, an I *think* that was getting a replacement BAM with the existing read groups plus the new read groups.  This would allow downstream analysis tools to maintain their assumptions about all data being in one BAM

Reply:
This is helpful @kbergin. Do you have any feeling for how likely it is that a submitter forgets a FASTQ. Feeling here was that we really wouldn't expect to see that very often.

Reply:
> Pipelines that do not require co-processing continue to subscribe to assay bundle events and are unaffected.

@diekhans @barkasn is this still true? From our conversation earlier this week, my understanding was that you could only create a subscription that receives all project submission events, unlike the existing subscription queries that allow you to match on a particular kind of data. This means we would only have one subscription that notifies us of every complete submission, and depending on the contents we would have to determine what pipeline to run. So if we received a project submission event consisting of SmartSeq2 data, but we still had a SmartSeq2-specific subscription we would trigger the analysis twice.

Reply:
While processing is still on a per-bundle basis, the input grouping needs to be passed as an output data group for the needs of the downstream consumers like Azul.

Thanks for point out the outdate text, it has been updated.